### PR TITLE
Implement C++ docs generation and split up docs by module

### LINKED ISSUE #22: Generate separate Sphinx files for each exposed module and an index file linking to them
Right now, all types in the environment are dumped into a single Sphinx file. We should emit one file per Diplomat module.

### PR DESCRIPTION
**Pull request by shadaj**
Example docs:
Screen Shot 2021-08-05 at 5 05 34 PM

Fixes #22

### CODE REVIEW

File: example/cpp/docs/decimal_ffi.rst [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+        The :cpp:class:`ICU4XFixedDecimalFormat`, exists if creation was successful.
+
+    .. cpp:member:: bool success
+
+
+        Whether creating the :cpp:class:`ICU4XFixedDecimalFormat` was successful.
+
+.. cpp:enum-struct:: ICU4XFixedDecimalGroupingStrategy
+
```
Comment (Manishearth) [inline example/cpp/docs/decimal_ffi.rst:48]: question: What's up with the extra whitespace?

File: tool/src/cpp/docs.rs [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+pub fn gen_method_docs<W: fmt::Write>(
+    out: &mut W,
+    method: &ast::Method,
+    in_path: &ast::Path,
+    env: &HashMap<ast::Path, HashMap<String, ast::ModSymbol>>,
+) -> fmt::Result {
+    let mut params: Vec<Param> = method.params.clone();
+    if method.is_writeable_out() {
```
Comment (Manishearth) [inline tool/src/cpp/docs.rs:132]: Probably should handle the templated Writeable out as well?

Comment (Manishearth) [reply tool/src/cpp/docs.rs:132]: Also, shouldn't this change the return type as well?

File: example/cpp/docs/decimal_ffi.rst (hunk 1 above)
Comment (shadaj) [reply example/cpp/docs/decimal_ffi.rst:48]: There aren't any docs, so you get a bunch of extra lines around where the docs would have been.

Comment (Manishearth) [reply example/cpp/docs/decimal_ffi.rst:48]: Ah. Would be nice if we can elide that

File: tool/src/cpp/docs.rs (hunk 2 above)
Comment (shadaj) [reply tool/src/cpp/docs.rs]: fixed!

File: tool/src/cpp/structs.rs [hunk 3]
```diff
@@ -309,6 +254,87 @@ fn gen_method<W: fmt::Write>(
     Ok(())
 }
 
+#[allow(clippy::too_many_arguments)]
+pub fn gen_method_interface<W: fmt::Write>(
```
Comment (Manishearth) [inline tool/src/cpp/structs.rs:258]: praise: good idea

File: example/cpp/docs/decimal_ffi.rst (hunk 1 above)
Comment (shadaj) [reply example/cpp/docs/decimal_ffi.rst:40]: fixed!

Comment (Manishearth) [reply example/cpp/docs/decimal_ffi.rst:39]: I think part of my issue is that at least for enum variants we should probably squish them all together. But it doesn't matter that much, this isn't human-read!

File: tool/src/docs_util.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+use pulldown_cmark::{BrokenLink, CowStr, Event, LinkType, Options, Parser, Tag};
+
+pub fn markdown_to_rst<W: fmt::Write>(
+    out: &mut W,
+    markdown: &str,
+    write_reference: &dyn Fn(&ast::Path, &mut W) -> fmt::Result,
+) -> fmt::Result {
+    let mut broken_link_callback = |broken: BrokenLink| {
+        Some((
+            CowStr::from(broken.reference.to_string()),
```
Comment (Manishearth) [inline tool/src/docs_util.rs:13]: I feel like we should just leave these broken, yes? Unless we're implementing intra doc links, which is a bunch of work

**Review by Manishearth — APPROVED** [review]

File: tool/src/cpp/docs.rs [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        .or_insert_with(String::new);
+    writeln!(index_out, "Documentation")?;
+    writeln!(index_out, "=============")?;
+    writeln!(index_out)?;
+    writeln!(index_out, ".. toctree::")?;
+    let mut toctree_indent = indented(index_out).with_str("   ");
+    writeln!(&mut toctree_indent, ":maxdepth: 3")?;
+    writeln!(&mut toctree_indent, ":caption: Modules:")?;
+    writeln!(&mut toctree_indent)?;
+    let mut sorted_keys: Vec<String> = env
```
Comment (Manishearth) [inline tool/src/cpp/docs.rs:28]: We really should provide the ability to name our modules

Comment (shadaj) [reply tool/src/cpp/docs.rs:28]: #66

File: tool/src/docs_util.rs (hunk 4 above)
Comment (shadaj) [reply tool/src/docs_util.rs:13]: We need this to support inter-doc links, since otherwise we don't get events for the link.

Comment (Manishearth) [reply tool/src/docs_util.rs:13]: Right, but we don't have that support yet, so is this doing anything? (I might me missing something?)

Comment (shadaj) [reply tool/src/docs_util.rs:13]: We do! You can see us set `in_shortcut` to true when we see the broken link and then generate the reference when we get an `Event::Code`.

Comment (Manishearth) [reply tool/src/docs_util.rs:13]: (anyway this is non blocking)

Comment (Manishearth) [reply tool/src/docs_util.rs:13]: But we should link to an issue about intradoc

Comment (shadaj) [reply tool/src/docs_util.rs:13]: Wait, hmm, not sure I understand. We _do_ support intra-doc links for types.

Comment (Manishearth) [reply tool/src/docs_util.rs:13]: _oh_, I see now.

**Review by Manishearth — APPROVED** [review]

**Review by sffc — COMMENTED** [review]
LGTM

**Pull request merged**